### DISCUSSION TITLE
How can i add extra functions to the different dialects

### DISCUSSION THREAD

**Discussion by WilliamStam** [sqlalchemy/sqlalchemy#11155] [Usage Questions] [answered]
It seems sqlalchemy2 has removed func.dateadd(). (google results all show to use that but its nowhere to be found in the package)

I found this https://github.com/sqlalchemy/sqlalchemy/issues/5672 here which says dateadd is a no go. so that leads me down another rabbit hole.. my use case is to `DATE_ADD(le_column, INTERVAL le_another_column_with_seconds SECOND) as new_val` (mysql dialect)

 Is it possible to define a set of functions and apply them to dialects so that i can do something like `func.date_add_seconds(date_column,seconds_column)` which will render as per the above? i'll need to implement this for the usual dialects (mysql,oracle,pg,sqlite) 

Really hoping there's an easy way to do something like 

```
def date_add_seconds():
  if dialect == x
    return ....
  elif dialect == y
    return ... 


func.add(date_add_seconds)
```

But since the selectable is created before the engine gets involved (to find the dialect) im guessing im SOoL on this 1

**Comment by zzzeek** [comment]
hi -

SQLAlchemy's SQL functions are dynamically generarated.  If you have a SQL function called `work_the_frobnozzle()`, here it is: `select(func.work_the_frobnozzle())`, and you got it.

The SQLAlchemy Tutorial has an in-depth section on functions: https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#working-with-sql-functions

**Reply by WilliamStam** [reply]
thanks for the response :D 

i noticed that when i did `func.something` it outputted it as `something()` - super cool :D - in the stmt but i couldn't find a way to do it "contitionaly" based on dialect. 

ie define a function like `date_add_seconds(<date column>, seconds)` that outputs:

mysql: `select date_add(date_column, INTERVAL 20 second) from table`
oracle: `select date_column + interval '20' second  from dual`
postgres: `select date_column + INTERVAL '1 day'`

**Reply by zzzeek** [reply, marked as answer]
for the conditional thing I think the approach at https://docs.sqlalchemy.org/en/20/core/compiler.html#utc-timestamp-function is more expedient

**Reply by WilliamStam** [reply]
ok... probs defs not the first person to say this and defs not the last.... but just plain straight "wow"

**Comment by WilliamStam** [comment]
```
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql import (
    expression,
)
from sqlalchemy.types import DateTime


class date_add_seconds(expression.FunctionElement):
    type = DateTime()
    name = 'date_add_seconds'
    inherit_cache = True


@compiles(date_add_seconds)
def default_date_add_seconds(element, compiler, **kw):
    arg1, arg2 = list(element.clauses)
    return "date_add(%s,interval %s second)" % (
        compiler.process(arg1, **kw),
        compiler.process(arg2, **kw)
    )


@compiles(date_add_seconds, 'sqlite')
def case_date_add_seconds(element, compiler, **kw):
    arg1, arg2 = list(element.clauses)
    return "datetime(%s, +%s seconds" % (
        compiler.process(arg1, **kw),
        compiler.process(arg2, **kw)
    )


@compiles(date_add_seconds, 'oracle')
@compiles(date_add_seconds, 'postgresql')
def case_date_add_seconds(element, compiler, **kw):
    arg1, arg2 = list(element.clauses)
    return "%s + interval %s second" % (
        compiler.process(arg1, **kw),
        compiler.process(arg2, **kw)
    )


```

usage

```
func.now() <= date_add_seconds(JobsView.last_ran,JobsView.seconds_between_run)
# or
func.now() <= date_add_seconds(JobsView.last_ran,text("40"))
```

**Reply by CaselIT** [reply]
use `compiler.process(arg1, **kw), compiler.process(arg2, **kw)` so you don't rely on the naive `__str__` of each element

**Reply by zzzeek** [reply]
great.  we have stayed away from the date thing because it's too huge to try to get all date arithmetic normalized across backends, we'd be fighting TZ / off by one issues for years

**Reply by WilliamStam** [reply]
i saw :D and i honestly dont blame you. making something "sooooo generic" is one of the hardest things to do. 

> use `compiler.process(arg1, **kw), compiler.process(arg2, **kw)` 

gotcha. i'm going to edit the above for anyone else landing up here
